src/docs/arch-locks.dox

```diff
@@ -0,0 +1,124 @@
+@row{\c log_slot_lock, Consolidation array}
+@row{\c log_sync_lock, Single-thread fsync}
+@row{\c log_writelsn_lock, Write LSN}
+@row{\c log_remove_lock, Remove and log cursors\, R/W Lock}
+@row{\c debug_log_retention_lock, Log retention reconfiguration lock\, R/W Lock}
+</table>
+
+<b>Locks hierarchy</b>:
+
+<table class=lock-hierarchy>
+@row{\c log_slot_lock, &rarr;, \c log_fs_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{\c log_slot_lock, &rarr;, \c log_writelsn_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{WRITE \c log_remove_lock, &rarr;, READ \c debug_log_retention_lock}
+</table>
+
+NOTE: Some functions expect \c log_slot_lock to be acquired, then they temporarily release it and re-acquire it.
+
+@section arch-locks-checkpoint-schema Checkpoint and Schema locks
+
+<table class=lock-hierarchy>
+@row{\c checkpoint_lock, &rarr;, \c schema_lock, &rarr;, WRITE \c table_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{\c checkpoint_lock, &rarr;, \c schema_lock, &rarr;, READ \c table_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{\c checkpoint_lock, &rarr;, \c schema_lock, &rarr;, \c api_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{\c checkpoint_lock, &rarr;, \c schema_lock, &rarr;, \c metadata_lock, &rarr;, READ \c table_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{\c checkpoint_lock, &rarr;, \c schema_lock, &rarr;, \c turtle_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{\c checkpoint_lock, &rarr;, \c schema_lock, &rarr;, WRITE \c dhandle_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{\c checkpoint_lock, &rarr;, \c turtle_lock}
+</table>
+
+NOTE: Checkpoint and schema locks are mostly used via \c WT_WITH_*_LOCK and the functions
+are covered with \c WT_ASSERT_SPINLOCK_OWNED statements which make the code easier to read.
+
+@section arch-locks-checkpoint-misc Miscellaneous locks
+
+<table class=lock-hierarchy>
+@row{WRITE \c table_lock, &rarr;, WRITE \c dhandle_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{\c reconfig_lock, &rarr;, WRITE \c chunkcache->pinned_objects.array_lock}
+</table>
+
+<table class=lock-hierarchy>
+@row{\c reconfig_lock, &rarr;, \c chunkcache->hashtable[*].bucket_lock, &rarr;, READ \c chunkcache->pinned_objects.array_lock}
+</table>
+
+@section arch-locks-examples Examples
+
+@subsection arch-locks-examples-eviction0 Eviction (no queue lock)
+
+@image html locks-evict-example-simple.svg width=900
+<!-- callgraph -f '///evict_walk_lock|evict_pass_lock|WT_WITH_PASS_LOCK' -t '///evict_walk_lock|evict_pass_lock|WT_WITH_PASS_LOCK' -t __evict_clear_walk -t __evict_walk_tree -continue -tagrex '/(__wt_spin_[^u]\w++)\({ARG},\s++({ARG})/index($2, "evict") != -1 && index($2, "queue") == -1 ? "$1($2)" : undef' -tag "WT_WITH_PASS_LOCK(evict_pass_lock)=WT_WITH_PASS_LOCK" -x /_curhs_cache -x /_verbose_dump_cache -x /_cache_destroy -x /_cache_create -d 11 -dir td -dot | dot -Tsvg /dev/stdin > src/docs/images/locks-evict-example-simple.svg -->
+
```

Comment:
Just confirming: this comment documents how the graph was created? Is callgraph in the code base?

Reply:
It's not in the code base but this comment will help to re-build it in future.